peer/peer.go

```diff
@@ -2043,29 +2069,60 @@ func (p *Peer) writeLocalVersionMsg() error {
 		return err
 	}
 
-	return p.writeMessage(localVerMsg, wire.LatestEncoding)
+	if err := p.writeMessage(localVerMsg, wire.LatestEncoding); err != nil {
+		return err
+	}
+
+	return nil
 }
 
-// negotiateInboundProtocol waits to receive a version message from the peer
-// then sends our version message. If the events do not occur in that order then
-// it returns an error.
+// negotiateInboundProtocol performs the negotiation protocol for an inbound
+// peer. The events should occur in the following order, otherwise an error is
+// returned:
+//
+//   1. Remote peer sends their version.
+//   2. We send our version.
+//   3. We send our verack.
+//   4. Remote peer sends their verack.
```

Comment:
We may want to add some grace to this ordering here, to ensure that both sides don't deadlock if we're connecting to an implementation with a slightly different ordering.

Reply:
If they send a message in the wrong order, then an error would be returned which would cause us to disconnect them. If they don't sent a message we expect, there's a timeout that'll also cause us to disconnect them. I believe that should cover everything?

Reply:
Yeah it's also already the case that we have an expected order as is.